### PR TITLE
ARC-1643 : add outbound proxy to config class

### PR DESCRIPTION
**Pull request by bgvozdev**
**What's in this PR?**
I'm adding ability to change the list of GHES testing hostnames/IPs via a feature flag.

**Why**
1. Current implementation skips out proxy for calls to all hostnames ending ".atlassian.com", which is too broad
2. Current implementation doesn't allow to add more hostnames to skip on the fly (e.g. for a pollinator, or while testing with your own GHES instance)
3. Currently the outbound proxy code is inside github-client, which makes it impossible to reuse on the "Connect GHE server" page

**Added feature flags**
- `use-outbound-proxy-skiplist` - the boolean flag for rollout of the feature
- `outbound-proxy-skiplist` - a permanent flag to serve with the list of internal Atlassian instances

**Affected issues**  
ARC-1643

**How has this been tested?**  
Manual testing locally and in ddev

**Whats Next?**
Making client on "Connect GHE server" page to reuse the new code

### CODE REVIEW

**Review by gxueatlassian — APPROVED** [review]
LGTM, maybe wait for another approval?

File: src/util/get-github-client-config.ts
Comment (bgvozdev) [reply src/util/get-github-client-config.ts:50]: Sorry, @gxueatlassian , could you please elaborate what do you mean by "above"? About the whole block or some particular logic/line?

Comment (gxueatlassian) [reply src/util/get-github-client-config.ts:50]: I mean why do we need to `hostname => hostname.trim()` or `hostname => new URL(hostname).host.toLowerCase()` etc? As those value comes from the feature flags, we can just always put it as exact values we like, such as `https://always-lower-case.com,https://no-space-around-comma.com`

Comment (bgvozdev) [reply src/util/get-github-client-config.ts:50]: Yeah, true, I was just trying to be a little bit more resilient against these subtle mistakes... Also, the name of the feature flag doesn't specify what exactly is expected (it is not `outbound-proxy-baseurl-skiplist`), so I guess this logic is kinda needed, otherwise we also need to rename the FF (to remove the 2nd  and 3rd maps). If we want to get rid of the 1st map, I think it is safer to convert the value into JSON in such case, so all the strings are wrapped with quotes and it is easy to spot such mistake.

Please let me know if you find this as some issue which you'd like to have been fixed or as a nit. I find the current solution OK but don't have a strong opinion and happy to adjust it if you have one ;)

Comment (gxueatlassian) [reply src/util/get-github-client-config.ts:50]: It is definitely /nit. 

However, I think the impact is super low ( if something messup in the list, the impact is we can't connect to our internal github server for testing/pollinator purpose ). All PROD use scenario should NOT skip the the proxy.

That's why I prefer avoid the code logic at all. If we didn't put the correct value in the lauchdarkly, we will find out (as it is internal github), can always correct it. Code can be simpler. 

But anyway, definitely nit, so you can leave it alone.

Comment (bgvozdev) [reply src/util/get-github-client-config.ts:50]: > if something messup in the list, the impact is we can't connect to our internal github server for testing/pollinator purpose

True, however debugging could take some time :)

Comment (gxueatlassian) [reply src/util/get-github-client-config.ts:50]: Right. All good mate, then leave it in :)

File: src/util/get-github-client-config.ts
Comment (mboudreau) [inline src/util/get-github-client-config.ts:70]: I am not a fan of this pattern as you're making the original config much harder to read and follow, and can also produce side effects.  Instead I'd recommend you do 
```
return {
   serverId: gitHubServerApp.id,
   ...
   proxyBaseUrl: this.setupProxyConfig(logger, jiraHost, gitHubServerApp.gitHubBaseUrl)
   };
```

Also, as a standard, we've been moving logger to the last argument in a function list as the other information is more important.

**Review by mboudreau — APPROVED** [review]

**Pull request merged**
